stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,97 @@
+    system such that no more than one mutable alias for the underlying data
+    may exist. Consider Boxed[T] over UnsafePointer[T] where possible.
+    """
+
+    var _inner: UnsafePointer[T, address_space]
+
+    fn __init__[T: Movable](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
+        """Construct a new Boxed[] by moving the passed value into a new backing allocation.
+
+        Args:
+            value: The value to move into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, AddressSpace.GENERIC].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[T: ExplicitlyCopyable, __:None=None](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_value: T):
+        """Construct a new Boxed[] by explicitly copying the passed value into a new backing allocation.
+
+        Args:
+            copy_value: The value to explicitly copy into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, address_space].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    fn __init__[T: ExplicitlyCopyable](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_box: Boxed[T, AddressSpace.GENERIC]):
+        """Construct a new Boxed[] by explicitly copying the value from another Boxed[].
+
+        Args:
+            copy_box: The Boxed[] to copy.
+        """
+        self.__init__(copy_value=copy_box[])
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Move this Boxed[].
+
+        Args:
+            existing: The value to move.
+        """
+        self._inner = existing._inner
+        existing._inner = UnsafePointer[T, address_space]()
+
+    fn __getitem__(ref [_, address_space._value.value] self) -> ref [__lifetime_of(self._inner), address_space._value.value] T:
+        """Returns a reference to the box's underlying data with parametric mutability.
+
+        Returns:
+            A reference to the data underlying the Boxed[].
+        """
+        # This should have a widening conversion here that allows
+        # the mutable ref that is always (potentially unsafely)
+        # returned from UnsafePointer to be guarded behind the
+        # aliasing guarantees of the lifetime system here.
+        # All of the magic happens above in the function signature
+        var inner_not_null = self._inner.__bool__()
+
+        debug_assert(inner_not_null, "Box is horribly broken, and __getitem__ was called on a destroyed box")
+
+        return self._inner[]
+
+    fn __del__(owned self: Boxed[T, AddressSpace.GENERIC]):
+        """Destroy the Boxed[]."""
+        # check that inner is non-null to accomodate into_inner and other
+        # consuming end states
+        if self._inner:
+            (self._inner).destroy_pointee()
+            self._inner.free()
+            self._inner = UnsafePointer[T, AddressSpace.GENERIC]()
+
+    fn into_inner[T: Movable](owned self: Boxed[T, AddressSpace.GENERIC]) -> T:
```

Comment:
```suggestion
    fn __getitem__[T: Movable](owned self: Boxed[T, AddressSpace.GENERIC]) -> T:
```
I think this aligns a bit more to `Reference` syntax i.e. `something_to_deref[]`

Reply:
into_inner was something I made explicitly separate from the regular `__getitem__`, it has consume semantics instead. There is still a `__getitem__` to get a ref to the data above this.

Reply:
Oh sorry didn't read with attention. This might make more sense as a `take()` function then

Reply:
I considered it, I'm not sure the semantics I'd associate with "take" make as much sense here, since it doesn't actually leave behind a "null box"--it fully consumes the original box, and requires moving it into the method. Maybe a `consume()`? I just took the Rust nomenclature for it and figured it was at least a reasonable first cut.

Reply:
I'd recommend calling the type `OwnedPointer` or just `Owned`, and then call the function `take()`, `take_data()`, `unbox()`, but naming isn't too important right now, we can figure it out later.

Reply:
Also, we might want to have an `unsafe_ptr()` method, (even though you can just access the field) it would follow the current stdlib conventions